### DISCUSSION TITLE
Install windows/amd64 binaries on windows/arm64

### DISCUSSION THREAD

**Discussion by CrystalMethod** [aquaproj/aqua#2514] [Deprecated (Feature Request)] [answered]
### Feature Overview

Windows 11 on arm64 is able to run binaries natively compiled for amd64. So Aqua should install amd64 binaries on arm64 platform if no arm64 binary is available.

### Why is the feature needed?

There are not so many tools that are compiled natively for windows/arm64. Supporting tools for windows/amd64 would significantly increase the visibility and usability of Aqua on windows/arm64. Same as Rosetta2.

EDIT:
https://learn.microsoft.com/en-us/windows/arm/apps-on-arm-x86-emulation

**Comment by suzuki-shunsuke** [comment]
Thank you for your report.
I'm not familiar with Windows, so I didn't know Windows's x64 emulation.

https://learn.microsoft.com/en-us/windows/arm/arm64ec

I don't have Windows arm64 machine so I can't verify aqua on that, but as you said maybe a feature like [rosetta2](https://aquaproj.github.io/docs/reference/registry-config/rosetta2/) for Windows is useful.

```yaml
windows_x64_emulation: true
```

**Reply by suzuki-shunsuke** [reply]
Hmm. Sorry. The reference may be unrelated. I can't find the reference about executing binaries for windows/amd64 on Windows arm64.

**Reply by CrystalMethod** [reply]
https://learn.microsoft.com/en-us/windows/arm/apps-on-arm-x86-emulation

**Reply by CrystalMethod** [reply]
```console
$ file scoop/apps/aqua/current/aqua.exe
scoop/apps/aqua/current/aqua.exe: PE32+ executable (console) Aarch64, for MS Windows, 6 sections

$ file scoop/apps/chezmoi/current/chezmoi.exe
scoop/apps/chezmoi/current/chezmoi.exe: PE32+ executable (console) Aarch64, for MS Windows, 7 sections

$ file scoop/apps/neovim/current/bin/nvim
scoop/apps/neovim/current/bin/nvim: PE32+ executable (console) x86-64, for MS Windows, 6 sections
```

and neovim runs totally fine.

**Reply by suzuki-shunsuke** [reply]
Do you have Windows ARM64 machines and you can execute binaries for windows/amd64 on those machines?

**Reply by CrystalMethod** [reply]
Yes, I recently switch from Intel based MacBook to M2 based Mac Studio. I need a Windows installation for various customer projects. Since I don't want to install any additional hardware, I currently have Windows 11 running on the M2 using Parallels. It runs surprisingly stable - even if it still only has "preview" status.

**Reply by suzuki-shunsuke** [reply]
https://learn.microsoft.com/en-us/windows/arm/apps-on-arm-x86-emulation

> Emulation for x86 apps makes the rich ecosystem of Win32 apps available on Arm. This provides the user the magical experience of running an existing x86 win32 app without any modifications to the app.

I think this is the emulation to execute Win32 apps, not amd64.

**Reply by suzuki-shunsuke** [reply]
> Yes, I recently switch from Intel based MacBook to M2 based Mac Studio. I need a Windows installation for various customer projects. Since I don't want to install any additional hardware, I currently have Windows 11 running on the M2 using Parallels. It runs surprisingly stable - even if it still only has "preview" status.

Oh, I see. Thank you.

**Reply by CrystalMethod** [reply]
```console
$ scoop list
Installed apps:

Name    Version                  Source   Updated             Info
----    -------                  ------   -------             ----
7zip    23.01                    main     2023-11-23 09:11:30
aqua    2.19.0                   aquaproj 2023-11-23 08:57:02
chezmoi 2.41.0                   main     2023-11-22 23:08:07
git     2.43.0                   main     2023-11-23 08:28:42 64bit
msys2   2023-10-26               main     2023-11-23 09:18:56 64bit
neovim  0.9.4                    main     2023-11-23 09:04:44 64bit
pwsh    7.4.0                    main     2023-11-22 23:26:26
wezterm 20230712-072601-f4abf8fd extras   2023-11-22 23:47:20 64bit
```

While 7zip, aqua, chezmoi and pwsh are native arm64 binaries - the others are x86_64 binaries.

> Windows 10 include emulation technology that enables existing unmodified x86 apps to run on Arm devices. Windows 11 extends that emulation to run unmodified x64 Windows apps on Arm-powered devices.

**Reply by suzuki-shunsuke** [reply]
Oh, thank you.

https://learn.microsoft.com/en-us/windows/arm/add-arm-support#emulation-on-arm-based-devices-for-x86-or-x64-windows-apps

> Windows 11 extends that emulation to run unmodified x64 Windows apps on Arm-powered devices.

**Reply by suzuki-shunsuke** [reply]
Built v2.20.0-1. Could you try this?

- https://github.com/aquaproj/aqua/pull/2515
- https://github.com/aquaproj/aqua/releases/tag/v2.20.0-1

The field `windows_arm_emulation` was added.

e.g. https://github.com/aquaproj/aqua-registry/blob/dd1db2ede12b1147c9c97a2341500bd0dc0e4438/pkgs/FiloSottile/age/registry.yaml#L6

## How to test

Checkout https://github.com/aquaproj/aqua-registry/pull/17622 and test on windows/arm64.

```console
$ gh pr checkout 17622
$ aqua g -i FiloSottile/age
$ aqua i -l
$ aqua which age
$ age --help
```

**Reply by CrystalMethod** [reply]
awesome, looks very neat and promising..

```console
$ aqua i -l
INFO[0000] create a proxy file                           aqua_version=2.20.0-1 env=windows/arm64 program=aqua proxy_path="C:\\Users\\...\\AppData\\Local\\aquaproj-aqua\\bin\\age"
INFO[0000] create a proxy file                           aqua_version=2.20.0-1 env=windows/arm64 program=aqua proxy_path="C:\\Users\\...\\AppData\\Local\\aquaproj-aqua\\bat\\age.bat"
INFO[0000] create a proxy file                           aqua_version=2.20.0-1 env=windows/arm64 program=aqua proxy_path="C:\\Users\\...\\AppData\\Local\\aquaproj-aqua\\bin\\age-keygen"
INFO[0000] create a proxy file                           aqua_version=2.20.0-1 env=windows/arm64 program=aqua proxy_path="C:\\Users\\...\\AppData\\Local\\aquaproj-aqua\\bat\\age-keygen.bat"

$ aqua which age
C:\Users\...\AppData\Local\aquaproj-aqua\pkgs\github_release\github.com\FiloSottile\age\v1.1.1\age-v1.1.1-windows-amd64.zip\age\age.exe

$ age --help
INFO[0000] download and unarchive the package            aqua_version=2.20.0-1 env=windows/arm64 exe_name=age package_name=FiloSottile/age package_version=v1.1.1 program=aqua registry=standard
WARN[0000] remove a temporal file                        aqua_version=2.20.0-1 env=windows/arm64 error="remove C:\\Users\\..,\\scoop\\apps\\msys2\\2023-10-26\\tmp\\011551611: Der Prozess kann nicht auf die Datei zugreifen, da sie von einem anderen Prozess verwendet wird." exe_name=age package_name=FiloSottile/age package_version=v1.1.1 program=aqua registry=standard
Usage:
    age [--encrypt] (-r RECIPIENT | -R PATH)... [--armor] [-o OUTPUT] [INPUT]
    age [--encrypt] --passphrase [--armor] [-o OUTPUT] [INPUT]
    age --decrypt [-i PATH]... [-o OUTPUT] [INPUT]

Options:
    -e, --encrypt               Encrypt the input to the output. Default if omitted.
    -d, --decrypt               Decrypt the input to the output.
    -o, --output OUTPUT         Write the result to the file at path OUTPUT.
    -a, --armor                 Encrypt to a PEM encoded format.
    -p, --passphrase            Encrypt with a passphrase.
    -r, --recipient RECIPIENT   Encrypt to the specified RECIPIENT. Can be repeated.
    -R, --recipients-file PATH  Encrypt to recipients listed at PATH. Can be repeated.
    -i, --identity PATH         Use the identity file at PATH. Can be repeated.

INPUT defaults to standard input, and OUTPUT defaults to standard output.
If OUTPUT exists, it will be overwritten.

RECIPIENT can be an age public key generated by age-keygen ("age1...")
or an SSH public key ("ssh-ed25519 AAAA...", "ssh-rsa AAAA...").

Recipient files contain one or more recipients, one per line. Empty lines
and lines starting with "#" are ignored as comments. "-" may be used to
read recipients from standard input.

Identity files contain one or more secret keys ("AGE-SECRET-KEY-1..."),
one per line, or an SSH key. Empty lines and lines starting with "#" are
ignored as comments. Passphrase encrypted age files can be used as
identity files. Multiple key files can be provided, and any unused ones
will be ignored. "-" may be used to read identities from standard input.

When --encrypt is specified explicitly, -i can also be used to encrypt to an
identity file symmetrically, instead or in addition to normal recipients.

Example:
    $ age-keygen -o key.txt
    Public key: age1ql3z7hjy54pw3hyww5ayyfg7zqgvc7w3j2elw8zmrj2kg5sfn9aqmcac8p
    $ tar cvz ~/data | age -r age1ql3z7hjy54pw3hyww5ayyfg7zqgvc7w3j2elw8zmrj2kg5sfn9aqmcac8p > data.tar.gz.age
    $ age --decrypt -i key.txt -o data.tar.gz data.tar.gz.age

$ which age
/c/Users/.../AppData/Local/aquaproj-aqua/bin/age

$ file "$(aqua which age)"
C:\Users\...\AppData\Local\aquaproj-aqua\pkgs\github_release\github.com\FiloSottile\age\v1.1.1\age-v1.1.1-windows-amd64.zip\age\age.exe: PE32+ executable (console) x86-64 (stripped to external PDB), for MS Windows, 13 sections```

**Reply by suzuki-shunsuke** [reply, marked as answer]
aqua v2.20.0 is out 🎉

https://github.com/aquaproj/aqua/releases/tag/v2.20.0

A field `windows_arm_emulation` was added.

**Comment by CrystalMethod** [comment]
There is just one really strange behavior: I would say if someone uses the `aqua-installer`, then everything would be fine - even on arm64 - with `< 2.20.0`, except that `aqua` will **only install amd64** binaries. This is because `aqua-installer` doesn't recognize the `arm64` arch but `amd64` because `uname -m` reports `x86_64` because the core msys2 installation ist still x86_64.

```console
$ aqua info # installed using aqua-installer
{
  "version": "2.20.0",
  "commit_hash": "61aaaa971b1fc06ff782b458e4ff85886fb61f39",
  "os": "windows",
  "arch": "amd64",
  "pwd": "C:\\Users\\...\\.local\\share\\chezmoi",
  "root_dir": "C:/Users/.../AppData/Local/aquaproj-aqua/",
  "env": {
    "AQUA_GLOBAL_CONFIG": "C:/Users/.../.config/aqua/aqua.yaml",
    "AQUA_POLICY_CONFIG": "C:/Users/.../.config/aqua/aqua-policy.yaml",
    "AQUA_ROOT_DIR": "C:/Users/.../AppData/Local/aquaproj-aqua/"
  },
  "config_files": []
}
```

While the manually downloaded pre-release correctly reports `arm64`

```console
$ aqua info # manually download aqua for arm64
{
  "version": "2.20.0-1",
  "commit_hash": "1a20f34a0560d5e07147745376c251b2257cfed4",
  "os": "windows",
  "arch": "arm64",
  "pwd": "C:\\Users\\...\\.local\\share\\chezmoi",
  "root_dir": "C:/Users/.../AppData/Local/aquaproj-aqua/",
  "env": {
    "AQUA_GLOBAL_CONFIG": "C:/Users/.../.config/aqua/aqua.yaml",
    "AQUA_POLICY_CONFIG": "C:/Users/.../.config/aqua/aqua-policy.yaml",
    "AQUA_ROOT_DIR": "C:/Users/.../AppData/Local/aquaproj-aqua/"
  },
  "config_files": []
}
```

I can also ask `scoop` to install a package for a specific architecture.

```console
$ scoop install aqua --arch <32bit|64bit|arm64>
```
But AFAIK `scoop` uses Powershell ps1 to detect os and arch. So

```console
$ scoop install aqua
```
will install the arm64 binary even if started from msys2.

Could it be a solution to modify `aqua-installer` to give the arch to the installer script and only fallback to `uname` if no arch is given?

```console
$ curl -sSfL https://raw.githubusercontent.com/aquaproj/aqua-installer/v2.2.0/aqua-installer | bash -s -- -v v2.20.0 -a arm64
```

**Reply by suzuki-shunsuke** [reply]
It looks weird.

https://github.com/aquaproj/aqua-installer/blob/736b7654e5031514207dac12b8d8e977f5fa3777/aqua-installer

aqua-installer v2.2.0 does the fowlloing things.

1. Download aqua v2.16.4 (aqua v2.16.4 is used for bootstrapping)
2. Execute `aqua update-aqua` with aqua v2.16.4 to install aqua

This means aqua-installer installs the latest version with aqua v2.16.4, so the change of aqua v2.20.0 is unrelated and aqua-installer should install aqua windows/arm64 on windows/arm64 correctly. 🤔

I don't have the machine windows/arm64, so I can't verify unfortunately.

**Reply by suzuki-shunsuke** [reply]
```console
$ aqua info # installed using aqua-installer
{
  "version": "2.20.0",
  "commit_hash": "61aaaa971b1fc06ff782b458e4ff85886fb61f39",
  "os": "windows",
  "arch": "amd64",
```

Oh? According to the result of `aqua info`, `arch` is not `arm64` but `amd64`.

This value has nothing to do with the architecture of aqua.
aqua gets the value with [runtime.GOARCH](https://pkg.go.dev/runtime#pkg-constants).

https://github.com/aquaproj/aqua/blob/61aaaa971b1fc06ff782b458e4ff85886fb61f39/pkg/controller/info/info.go#L73

**Reply by suzuki-shunsuke** [reply]
> uname -m reports x86_64 because the core msys2 installation ist still x86_64.

Hmm. I'm not familiar with msys2, but this is an issue of msys2, right?

Maybe setting the environment variable `AQUA_GOARCH=arm64` is helpful.

```console
$ export AQUA_GOARCH=arm64
```

**Reply by suzuki-shunsuke** [reply]
Or if you can execute x64 binaries on windows/arm64, maybe you don't have to be worried about this issue.

**Comment by CrystalMethod** [comment]
Here is an open issue regarding `uname` on `arm64`: https://github.com/msys2/msys2-runtime/issues/171

Yes, with
```shell
export AQUA_GOARCH=arm64
```
I get `aqua/arm64@2.20.0` installed by `aqua-installer` and now `aqua i -a` installs all binaries available for `arm64`. But this is max the half of the tools given in `AQUA_GLOBAL_CONFIG`. So next I need to have the new registry and attribute `windows_arm_emulation` to get the rest of tools installed as amd64 binaries. And so I can benefit from the arm architecture if available but fallback to amd.

**Reply by suzuki-shunsuke** [reply]
Thank you for sharing the issue of msys2.
Yeah, we need to fix [the standard Registry (aqua-registry)](https://github.com/aquaproj/aqua-registry) using `windows_arm_emulation`. I'll write scripts for it when I have time.
